### PR TITLE
storage/v2: expose Gap object

### PR DESCRIPTION
**Pull request by dbungert**

### CODE REVIEW

File: subiquity/common/serialize.py [hunk 1]
```diff
@@ -194,6 +194,10 @@ def _deserialize_attr(self, annotation, value, metadata, path):
             args = {}
             fields = {field.name: field for field in attr.fields(annotation)}
             for key in value.keys():
+                if key not in fields and key == '$type':
+                    # Union types can contain a '$type' field that is not
+                    # actually one of the keys.
+                    continue
```
Comment (mwhudson) [inline subiquity/common/serialize.py:200]: Ugh is this a generic problem with List[Union]? :( Sounds like that can be fixed in a separate branch (with tests <wink>)

Comment (dbungert) [reply subiquity/common/serialize.py:200]: https://github.com/canonical/subiquity/pull/1177

**Review by mwhudson — APPROVED** [review]

File: subiquity/common/types.py [hunk 2]
```diff
@@ -268,14 +268,20 @@ class Partition:
     os: Optional[OsProber] = None
 
 
+@attr.s(auto_attribs=True)
+class Gap:
+    offset: int
+    size: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/common/types.py:274]: Perhaps it made sense at the time?  Fixing.

File: subiquity/tests/api/test_api.py [hunk 3]
```diff
@@ -735,6 +754,107 @@ async def test_win10(self):
             self.assertEqual(expected, sda1['os'])
 
 
+class TestRepartition(TestAPI):
```
Comment (mwhudson) [inline subiquity/tests/api/test_api.py:757]: I'm not sure "repartition" is the right jargon here. TestPartitionTableEditing?

File: subiquity/tests/api/test_api.py [hunk 4]
```diff
@@ -735,6 +754,107 @@ async def test_win10(self):
...
+            await inst.post('/storage/v2/add_partition', data)
+
+            # Adding a second should fail.
+            with self.assertRaises(ClientResponseError, msg=str(data)):
+                data['partition']['mount'] = '/usr'
+                await inst.post('/storage/v2/add_partition', data)
+
+    @timeout()
+    async def SKIP_test_try_to_use_free_space(self):
+        # disabled until we can repartition
```
Comment (mwhudson) [inline subiquity/tests/api/test_api.py:777]: ditto about jargon choices

**Review by mwhudson — APPROVED** [review]
Nice. Minor quibble about terminology.

I like that you have integration tests ready to be enabled when we get a bit further with this.

File: subiquity/tests/api/test_api.py (hunk 3 above)
Comment (dbungert) [reply subiquity/tests/api/test_api.py:757]: Fixed, thanks!

**Pull request merged**